File: spollex/doctype_events/sales_order.py
# Copyright (c) 2026, 4C Solutions and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def validate_subscription_details(doc, method=None):
    target_dt = "Sales Order" if getattr(doc, "custom_export", 0) else "Sales Invoice"
    doc.custom_old_doctype = target_dt

    is_new = getattr(doc, "custom_is_new_subscription", 0)
    if doc.custom_subscription_type in ["AMC", "License"] and not is_new and not doc.custom_old_invoice:
        label = _("Old Proforma Invoice") if getattr(doc, "custom_export", 0) else _("Old Invoice")
        frappe.throw(_("{0} is mandatory when Subscription Type is {1}").format(label, doc.custom_subscription_type))

def update_renewal_invoice_status(doc, method=None):
    old_invoice = getattr(doc, "custom_old_invoice", None)
    if not old_invoice:
        return

    if getattr(doc, "custom_export", 0):
        is_renewed = 1 if method == "on_submit" else 0
        target_dt = getattr(doc, "custom_old_doctype", None) or "Sales Order"
        if frappe.db.exists(target_dt, old_invoice):
            frappe.db.set_value(target_dt, old_invoice, "custom_renewed", is_renewed)

File: spollex/public/js/sales_order.js
frappe.ui.form.on("Sales Order", {
    onload: function(frm) {
        update_old_invoice_link(frm);
        frm.set_query("item_code", "custom_subscription_items", function(doc, cdt, cdn) {
            let items = (doc.items || []).map(item => item.item_code).filter(Boolean);
            return {
                filters: [
                    ["Item", "name", "in", items]
                ]
            };
        });
    },
    refresh: function(frm) {
        update_old_invoice_link(frm);
    },
    custom_export: function(frm) {
        update_old_invoice_link(frm);
        if (frm.doc.custom_old_invoice) {
            frm.set_value("custom_old_invoice", "");
        }
    },
    custom_subscription_type: function(frm) {
        update_old_invoice_link(frm);
    },
    custom_is_new_subscription: function(frm) {
        update_old_invoice_link(frm);
    },
    before_save: function(frm) {
        update_old_invoice_link(frm);
        let subscription_items = frm.doc.custom_subscription_items || [];
        let doc_items = frm.doc.items || [];

        let item_amount_map = {};
        doc_items.forEach(item => {
            if (item.item_code) {
                item_amount_map[item.item_code] = (item_amount_map[item.item_code] || 0) + (item.amount || 0);
            }
        });

        subscription_items.forEach(row => {
            if (row.item_code && item_amount_map[row.item_code] !== undefined) {
                frappe.model.set_value(row.doctype, row.name, "amount", item_amount_map[row.item_code]);
            }
        });
    }
});

function update_old_invoice_link(frm) {
    let target_dt = frm.doc.custom_export ? "Sales Order" : "Sales Invoice";
    let target_label = frm.doc.custom_export ? __("Old Proforma Invoice") : __("Old Invoice");
    if (frm.doc.custom_old_doctype !== target_dt) {
        frm.set_value("custom_old_doctype", target_dt);
    }
    frm.set_df_property("custom_old_invoice", "label", target_label);
}

frappe.ui.form.on("Subscription Item", {
    item_code: function(frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.item_code) {
            setTimeout(() => {
                let matched_item = (frm.doc.items || []).find(item => item.item_code === row.item_code);
                if (matched_item) {
                    frappe.model.set_value(cdt, cdn, "amount", matched_item.amount);
                } else {
                    frappe.msgprint(__("Item 0 is not present in the items table.", [row.item_code]));
                }
            }, 100);
        }
    }
});
